File: problems/daily/Java/20. Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        StringBuilder tent = new StringBuilder();
        for(int i = 0; i < s.length(); i++){
            int l = tent.length();

            if(s.charAt(i) == ')' && l > 0){
                if(tent.charAt(l-1) == '('){
                    tent.delete(l-1,l);
                }else {
                    return false;
                }

            }else if(s.charAt(i) == '}' && l > 0){
                if(tent.charAt(l-1) == '{'){
                    tent.delete(l-1,l);
                }else {
                    return false;
                }

            }else if(s.charAt(i) == ']' && l >0){
                if(tent.charAt(l-1) == '['){
                    tent.delete(l-1,l);
                }else {
                    return false;
                }
                
            }else{
                tent.append(s.charAt(i));
            }
        }
        return tent.length() == 0;
    }
}

File: problems/daily/Python3/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        kind = set(['(','{','['])
        pair = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        for i in range(len(s)):
            if s[i] in kind:
                stack.append(s[i])
            else:
                if stack and pair[s[i]] == stack[-1]:
                    stack.pop()
                else:
                    return False
        return len(stack) == 0